### PR TITLE
Nheko room directory

### LINKED ISSUE #328: Room directory
Nheko should allow browsing rooms in the room directory as well as joining them.

### PR DESCRIPTION
**Pull request by kamathmanu**
Closes #328 with an initial version of the Room Directory. This branch is the completed/restructured changes of my [original PR] (https://github.com/Nheko-Reborn/nheko/pull/496) since that one has some issues due to bad auto-merges.

**Features planned for room directory extension**

1. Including ability for users to specify server and network/protocols to search on (ideally with a dropdown/text-input combo UI)
2. Aesthetics (particularly to make the room information like number of members more intuitive/easily understandable)
3. Preview-ability

### CODE REVIEW

File: src/timeline/TimelineViewManager.cpp [hunk 1]
```diff
@@ -282,6 +286,9 @@ TimelineViewManager::TimelineViewManager(CallManager *callManager, ChatPage *par
                                          "EmojiCategory",
                                          "Error: Only enums");
 
+        qmlRegisterType<RoomDirectoryModel>(
+          "im.nheko.RoomDirectoryModel", 1, 0, "RoomDirectoryModel");
```
Comment (LorenDB) [inline src/timeline/TimelineViewManager.cpp:290]: Maybe just put it under `im.nheko`?

File: src/RoomDirectoryModel.h [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+        inline int rowCount(const QModelIndex &parent = QModelIndex()) const override
+        {
+                (void)parent;
+                return static_cast<int>(publicRoomsData_.size());
+        }
+
+        inline bool canFetchMore(const QModelIndex &) const override
+        {
+                nhlog::net()->debug("determining if can fetch more");
```
Comment (LorenDB) [inline src/RoomDirectoryModel.h:54]: Maybe this is a bit overkill for production? Maybe not? It seems like something that could clog up output, but maybe I'm being too picky

File: src/RoomDirectoryModel.h [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+                MemberCount,
+                Previewable
+        };
+        QHash<int, QByteArray> roleNames() const override;
+
+        QVariant data(const QModelIndex &index, int role) const override;
+
+        inline int rowCount(const QModelIndex &parent = QModelIndex()) const override
+        {
+                (void)parent;
```
Comment (LorenDB) [inline src/RoomDirectoryModel.h:48]: Nitpick: you could use `Q_UNUSED` instead, but it's not a big deal

File: src/RoomDirectoryModel.cpp [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+std::vector<std::string>
+RoomDirectoryModel::getViasForRoom(const std::vector<std::string> &aliases)
+{
+        std::vector<std::string> vias;
+
+        vias.reserve(aliases.size());
+
+        std::transform(
+          aliases.begin(), aliases.end(), std::back_inserter(vias), [](const auto &alias) {
+                  const auto roomAliasDelimiter = ":";
```
Comment (LorenDB) [inline src/RoomDirectoryModel.cpp:90]: Probably a one-use var like this is not necessary

File: src/RoomDirectoryModel.cpp [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+{
+        beginResetModel();
+
+        prevBatch_    = "";
+        nextBatch_    = "";
+        canFetchMore_ = true;
+
+        beginRemoveRows(QModelIndex(), 0, static_cast<int>(publicRoomsData_.size()));
+        publicRoomsData_.clear();
+        endRemoveRows();
```
Comment (LorenDB) [inline src/RoomDirectoryModel.cpp:47]: I don't think the remove rows calls are necessary if you are resetting the model

File: resources/res.qrc [hunk 6]
```diff
@@ -142,6 +142,12 @@
         <file>qml/emoji/EmojiPicker.qml</file>
         <file>qml/emoji/StickerPicker.qml</file>
         <file>qml/UserProfile.qml</file>
+        <file>qml/RoomDirectory.qml</file>
+	      <file>qml/delegates/MessageDelegate.qml</file>
+        <file>qml/delegates/TextMessage.qml</file>
+        <file>qml/delegates/NoticeMessage.qml</file>
+        <file>qml/delegates/ImageMessage.qml</file>
+        <file>qml/delegates/PlayableMediaMessage.qml</file>
```
Comment (LorenDB) [inline resources/res.qrc:150]: Why are these added? They exist elsewhere, and now building the resource file warns of duplicate aliases

File: resources/qml/RoomList.qml [hunk 7]
```diff
@@ -557,6 +564,11 @@ Page {
                     ToolTip.visible: hovered
                     ToolTip.text: qsTr("Room directory")
                     Layout.margins: Nheko.paddingMedium
+			onClicked: {
+                        console.debug("Roomdir clicked");
```
Comment (LorenDB) [inline resources/qml/RoomList.qml:568]: Maybe delete this :D Seems a little too verbose for production code

**Review by LorenDB — COMMENTED** [review]
N.B. I didn't mark all of the `Math.ceil` instances in here.

I'd also suggest a loading spinner to show the user that rooms are loading. Other than that, I like it.

File: resources/qml/RoomDirectory.qml [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+
+        implicitHeight: roomSearch.height
+
+        MatrixTextField {
+            id: roomSearch
+
+            Layout.fillWidth: true
+
+            font.pixelSize: fontMetrics.font.pixelSize
+            padding: Math.ceil(1.5 * Nheko.paddingSmall)
```
[the remark this answers is not in the document]
Comment (deepbluev7) [reply resources/qml/RoomDirectory.qml:43]: It depends on the value of the padding. Currently it would be fine, because the padding is 4 and 1.5*4 = 6. But if we change it to 5, it would misalign and break font rendering on most systems. But really, this should probably be paddingMedium...

File: resources/qml/RoomDirectory.qml [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+        model: RoomDirectoryModel {
+            id: roomDir
+        }
+        delegate: Rectangle {
+            id: roomDirDelegate
+
+            property color background: Nheko.colors.window
+            property color importantText: Nheko.colors.text
+            property color unimportantText: Nheko.colors.buttonText
+            property int avatarSize: Math.ceil(fontMetrics.lineSpacing * 4)
```
[the remark this answers is not in the document]
Comment (deepbluev7) [reply resources/qml/RoomDirectory.qml:70]: It is not needed if you multiply by an integer factor.

File: src/RoomDirectoryModel.h (hunk 3 above)
Comment (deepbluev7) [reply src/RoomDirectoryModel.h:48]: We use (void) in most places. Q_UNUSED is Qt specific and I personally don't think it is more readable. Alternatively there is also [[maybe_unused]]

File: resources/qml/RoomDirectory.qml [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+                        Layout.preferredWidth: parent.width
+                        spacing: Nheko.paddingSmall
+                        Layout.alignment: Qt.AlignVCenter | Qt.AlignLeft
+			Layout.preferredHeight: Math.ceil(fontMetrics.lineSpacing * 4)
+
+                        Label {
+                            id: roomTopic
+                            color: roomDirDelegate.unimportantText
+                            font.weight: Font.Thin
+			    Layout.alignment: Qt.AlignVCenter | Qt.AlignLeft
```
[the remark this answers is not in the document]
Comment (kamathmanu) [reply resources/qml/RoomDirectory.qml:130]: yeah i'll run it after incorporating your feedback

File: src/RoomDirectoryModel.h (hunk 3 above)
Comment (kamathmanu) [reply src/RoomDirectoryModel.h:48]: Yup I matched the convention to use void. [[maybe_unused]] would make Modern C++ fanatics happier but it seems a bit verbose for me in this context

File: src/RoomDirectoryModel.cpp (hunk 5 above)
Comment (kamathmanu) [reply src/RoomDirectoryModel.cpp:47]: @deepbluev7 Can you confirm this?

Comment (deepbluev7) [reply src/RoomDirectoryModel.cpp:47]: I can and I told you that before.

**Review by LorenDB — APPROVED** [review]
Looks fine to me

File: resources/qml/RoomDirectory.qml [hunk 11]
```diff
@@ -0,0 +1,216 @@
...
+import im.nheko 1.0
+
+ApplicationWindow {
+    id: roomDirectoryWindow
+
+    property RoomDirectoryModel publicRooms
+
+    visible: true
+    x: MainWindow.x + (MainWindow.width / 2) - (width / 2)
+    y: MainWindow.y + (MainWindow.height / 2) - (height / 2)
```
Comment (deepbluev7) [inline resources/qml/RoomDirectory.qml:18]:
```suggestion
    Component.onCompleted: Nheko.reparent(roomMembersRoot)
```

File: resources/qml/RoomDirectory.qml [hunk 12]
```diff
@@ -0,0 +1,216 @@
...
+        delegate: Rectangle {
+            id: roomDirDelegate
+
+            property color background: Nheko.colors.window
+            property color importantText: Nheko.colors.text
+            property color unimportantText: Nheko.colors.buttonText
+            property int avatarSize: fontMetrics.lineSpacing * 4
+
+            color: background
+            height: avatarSize + 2.5 * Nheko.paddingMedium
```
Comment (deepbluev7) [inline resources/qml/RoomDirectory.qml:47]: Please just use Nheko.paddingLarge instead of multiplying by random amounts.

File: resources/qml/RoomDirectory.qml [hunk 13]
```diff
@@ -0,0 +1,216 @@
...
+                            Layout.margins: Nheko.paddingSmall
+                            width: roomCount.width
+
+                            Label {
+                                id: roomCount
+
+                                color: roomDirDelegate.unimportantText
+                                anchors.centerIn: parent
+                                Layout.fillWidth: false
+                                font.weight: Font.Thin
```
Comment (deepbluev7) [inline resources/qml/RoomDirectory.qml:129]: Thin doesn't really work well on some systems, I would remove that.

```suggestion
```

File: resources/qml/RoomDirectory.qml [hunk 14]
```diff
@@ -0,0 +1,216 @@
...
+                            Layout.fillWidth: true
+                            text: model.topic
+                            verticalAlignment: Text.AlignVCenter
+                            wrapMode: Text.WordWrap
+                        }
+
+                        Item {
+                            id: numMembersRectangle
+
+                            Layout.fillWidth: false
```
Comment (deepbluev7) [inline resources/qml/RoomDirectory.qml:119]: You don't need to set that, false is the default. (Applies in multiple places)

File: resources/qml/RoomDirectory.qml [hunk 15]
```diff
@@ -0,0 +1,216 @@
...
+            anchors.margins: Nheko.paddingLarge
+
+            Spinner {
+                id: loadingSpinner
+
+                anchors.centerIn: parent
+                anchors.margins: Nheko.paddingLarge
+                running: visible
+                foreground: Nheko.colors.mid
+                z: 7
```
Comment (deepbluev7) [inline resources/qml/RoomDirectory.qml:178]: 7?

File: src/RoomDirectoryModel.cpp [hunk 16]
```diff
@@ -0,0 +1,196 @@
...
+        resetDisplayedData();
+}
+
+bool
+RoomDirectoryModel::canJoinRoom(const QByteArray &room)
+{
+        const auto &cache = cache::roomInfo();
+        const QString room_id(room);
+        const bool validRoom = !room_id.isNull() && !room_id.isEmpty();
+        return validRoom && !cache.contains(room_id);
```
Comment (deepbluev7) [inline src/RoomDirectoryModel.cpp:75]:
```suggestion
        return !room.isEmpty() && !cache::getRoomInfo(room.toStdString()).empty();

```

Comment (deepbluev7) [reply src/RoomDirectoryModel.cpp:75]: !isEmpty implies !isNull and you really don't need to load all rooms to check for a single rooms existence in the cache...

File: src/RoomDirectoryModel.h [hunk 17]
```diff
@@ -0,0 +1,93 @@
...
+
+signals:
+        void fetchedRoomsBatch(std::vector<mtx::responses::PublicRoomsChunk> rooms,
+                               const std::string &next_batch);
+        void loadingMoreRoomsChanged();
+        void reachedEndOfPaginationChanged();
+
+public slots:
+        void displayRooms(std::vector<mtx::responses::PublicRoomsChunk> rooms,
+                          const std::string &next_batch);
```
Comment (deepbluev7) [inline src/RoomDirectoryModel.h:75]: This could probably be a private slot

**Review by deepbluev7 — CHANGES_REQUESTED** [review]
It's getting there, but I have a small more cleanups for you ;-)

File: resources/qml/RoomDirectory.qml [hunk 18]
```diff
@@ -0,0 +1,216 @@
...
+    modality: Qt.WindowModal
+    flags: Qt.Dialog | Qt.WindowCloseButtonHint
+    title: qsTr("Explore Public Rooms")
+
+    Shortcut {
+        sequence: StandardKey.Cancel
+        onActivated: roomDirectoryWindow.close()
+    }
+
+    ListView {
```
Comment (deepbluev7) [inline resources/qml/RoomDirectory.qml:32]: Can you use the ScrollHelper in this? Otherwise it scrolls a bit fast on my system.

File: resources/qml/RoomDirectory.qml [hunk 19]
```diff
@@ -0,0 +1,215 @@
...
+
+                    RowLayout {
+                        id: roomNameRow
+
+                        spacing: 0
+
+                        ElidedLabel {
+                            Layout.alignment: Qt.AlignBottom
+                            color: roomDirDelegate.importantText
+                            elideWidth: textContent.width * 0.5 - Nheko.paddingMedium
```
Comment (deepbluev7) [inline resources/qml/RoomDirectory.qml:90]: This width seems arbitrary. Why can the room name only be half as wide as the column?

File: resources/qml/RoomDirectory.qml [hunk 20]
```diff
@@ -0,0 +1,215 @@
...
+                ColumnLayout {
+                    id: textContent
+
+                    Layout.alignment: Qt.AlignLeft
+                    width: parent.width - avatar.width
+                    Layout.preferredWidth: parent.width - avatar.width
+                    Layout.preferredHeight: roomNameRow.height + roomDescriptionRow.height
+                    spacing: Nheko.paddingSmall
+
+                    RowLayout {
```
Comment (deepbluev7) [inline resources/qml/RoomDirectory.qml:82]: You don't need rows with just one Element, do you?

**Comment by deepbluev7** [comment]
Looks good to me, thank you! And thank you for spending so much time on this and sticking through with it!
